File: apiGateway/controllers/followingControllers/followingController.js
const axios = require("axios");

async function createFollowing(req, res) {
  try {
    const result = await axios.post("http://127.0.0.1:5010/follow", req.body);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function unfollow(req, res) {
  try {
    const { followerID, followedID } = req.body;
    const result = await axios.delete("http://127.0.0.1:5010/unfollow", {
      headers: {
        "Content-Type": "application/json",
      },
      data: req.body,
    });
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getUserFollowers(req, res) {
  try {
    const followerID = req.params.followerID;
    const result = await axios.get(
      `http://127.0.0.1:5010/followings/followers/${followerID}`
    );
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getUserFollowed(req, res) {
  try {
    const followedID = req.params.followedID;
    const result = await axios.get(
      `http://127.0.0.1:5010/followings/followed/${followedID}`
    );
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function isFollowingExists(req, res) {
  try {
    const result = await axios.get("http://127.0.0.1:5010/exists", {
      headers: {
        "Content-Type": "application/json",
      },
      data: req.body, // Pass request body as data
    });
    res.json(result.data); // Send the response data
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

module.exports = {
  createFollowing,
  unfollow,
  getUserFollowers,
  getUserFollowed,
  isFollowingExists,
};

File: followingService/controllers/following_controller.py
from flask import Blueprint, jsonify, request
from repositories.following_repository import *

following_blueprint = Blueprint('following_blueprint', __name__)

@following_blueprint.route('/follow', methods=['POST'])
def create_following_controller():
    try:
        request_data = request.get_json()

        followerID = request_data.get("followerID")
        followedID = request_data.get("followedID")
        followDate = request_data.get("followDate")

        result = add_following_to_db(followerID, followedID, followDate)
      
        return jsonify({'result': result}), 201

    except Exception as e:
        return jsonify({'error': str(e)}), 500

@following_blueprint.route('/unfollow', methods=['DELETE'])
def unfollow_controller():
    try:
        request_data = request.get_json()

        followerID = request_data.get("followerID")
        followedID = request_data.get("followedID")
        result = unfollow(followerID,followedID)
        return jsonify({'result': result}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500

@following_blueprint.route('/followings/followers/<int:followerID>', methods=['GET'])
def get_user_followers_controller(followerID):
    try:
        following_list = get_user_followings(followerID)
        if following_list:
            return jsonify(following_list), 200
        else:
            return jsonify({'message': 'Followings not found'}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500
    
@following_blueprint.route('/followings/followed/<int:followedID>', methods=['GET'])
def get_user_followed_controller(followedID):
    try:
        following_list = get_user_followers(followedID)
        if following_list:
            return jsonify(following_list), 200
        else:
            return jsonify({'message': 'Followings not found'}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500

@following_blueprint.route('/exists', methods=['GET'])
def is_following_exists_controller():
    try:
        request_data = request.get_json()

        followerID = request_data.get("followerID")
        followedID = request_data.get("followedID")
        
        if followerID is None or followedID is None:
            return jsonify({'error': 'followerID and followedID are required parameters'}), 200
        
        exists = is_following_exists(followerID, followedID)
        return jsonify({'exists': exists}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500